lib/src/main/java/space/taran/arklib/domain/preview/PlainPreviewStorage.kt

```diff
@@ -14,29 +21,38 @@ import kotlin.io.path.createDirectories
 import kotlin.io.path.deleteIfExists
 import kotlin.io.path.isDirectory
 
-class PlainPreviewStorage(val root: Path) : PreviewStorage {
+class PlainPreviewStorage(
+    val root: Path,
+    private val index: PlainResourcesIndex,
+    private val appScope: CoroutineScope
+) : PreviewStorage {
     private val previewsDir = root.arkFolder().arkPreviews()
     private val thumbnailsDir = root.arkFolder().arkThumbnails()
 
     private fun previewPath(id: ResourceId): Path =
         previewsDir.resolve(id.toString())
+
     private fun thumbnailPath(id: ResourceId): Path =
         thumbnailsDir.resolve(id.toString())
 
     init {
         previewsDir.createDirectories()
         thumbnailsDir.createDirectories()
+        index.resourceDiffFlow.onEach { diff ->
+            appScope.launch(Dispatchers.IO) {
+                diff.added.forEach { (meta, path) ->
+                    launch { store(path, meta) }
+                }
+                diff.deleted.forEach { forget(it) }
+            }
+        }.launchIn(appScope + Dispatchers.IO)
+
     }
 
     override fun locate(path: Path, resource: ResourceMeta): PreviewAndThumbnail? {
         val thumbnail = thumbnailPath(resource.id)
         if (!Files.exists(thumbnail)) {
             Log.w(PREVIEWS, "thumbnail was not found for resource $resource")
-            if (Files.exists(previewPath(resource.id))) {
```

Comment:
If we throw away the assertion, then we should trigger thumbnail generation and not proceed with incorrect state.

Reply:
I think we need separate PR for this
Generating thumbnail right here is a bad idea because this method is called from main thread
We need queuing mechanism for this to avoid generating previews multiple times for same resource